File: 0001-0100/051-060/057M_Insert_Interval/java/main.java
// 57. Insert Interval
// https://leetcode.com/problems/insert-interval/
// T.C.: O(n)
// S.C.: O(1)

// java main.java

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

class Main {

  public static void main(String[] args) {
    Solution solution = new Solution();

    int[][] intervals = new int[][] { { 1, 3 }, { 6, 9 } };
    int[] newInterval = new int[] { 2, 5 };
    int[][] expected = new int[][] { { 1, 5 }, { 6, 9 } };
    int[][] result = solution.insert(intervals, newInterval);
    System.out.println(result + " " + (Arrays.deepEquals(expected, result) ? "true" : "false"));

    intervals = new int[][] { { 1, 2 }, { 3, 5 }, { 6, 7 }, { 8, 10 }, { 12, 16 } };
    newInterval = new int[] { 4, 8 };
    expected = new int[][] { { 1, 2 }, { 3, 10 }, { 12, 16 } };
    result = solution.insert(intervals, newInterval);
    System.out.println(result + " " + (Arrays.deepEquals(expected, result) ? "true" : "false"));

    intervals = new int[][] { { 1, 3 }, { 6, 9 } };
    newInterval = new int[] { 3, 5 };
    expected = new int[][] { { 1, 5 }, { 6, 9 } };
    result = solution.insert(intervals, newInterval);
    System.out.println(result + " " + (Arrays.deepEquals(expected, result) ? "true" : "false"));

    intervals = new int[][] { { 1, 3 }, { 6, 9 } };
    newInterval = new int[] { 9, 15 };
    expected = new int[][] { { 1, 3 }, { 6, 15 } };
    result = solution.insert(intervals, newInterval);
    System.out.println(result + " " + (Arrays.deepEquals(expected, result) ? "true" : "false"));

    intervals = new int[][] { { 1, 3 }, { 6, 9 } };
    newInterval = new int[] { 11, 15 };
    expected = new int[][] { { 1, 3 }, { 6, 9 }, { 11, 15 } };
    result = solution.insert(intervals, newInterval);
    System.out.println(result + " " + (Arrays.deepEquals(expected, result) ? "true" : "false"));
  }
}

class Solution {

  public int[][] insert(int[][] intervals, int[] newInterval) {
    List<int[]> result = new ArrayList<>();
    for (int i = 0; i < intervals.length; i++) {
      if (newInterval[1] < intervals[i][0]) {
        result.add(newInterval);
        for (int j = i; j < intervals.length; j++) {
          result.add(intervals[j]);
        }
        return result.toArray(new int[result.size()][]);
      } else if (newInterval[0] > intervals[i][1]) {
        result.add(intervals[i]);
      } else {
        newInterval[0] = Math.min(newInterval[0], intervals[i][0]);
        newInterval[1] = Math.max(newInterval[1], intervals[i][1]);
      }
    }
    result.add(newInterval);
    return result.toArray(new int[result.size()][]);
  }
}

File: 0001-0100/051-060/057M_Insert_Interval/rust/src/main.rs
// 57. Insert Interval
// https://leetcode.com/problems/insert-interval/
// T.C.: O(n)
// S.C.: O(1)
use colored::Colorize;
use colored::ColoredString;

struct Solution;

impl Solution {
    pub fn insert(intervals: Vec<Vec<i32>>, new_interval: Vec<i32>) -> Vec<Vec<i32>> {
        let mut result = vec![];
        let mut new_interval = new_interval;
        for (i, interval) in intervals.iter().enumerate() {
            if new_interval[0] > interval[1] {
                result.push(interval.clone());
            } else if new_interval[1] < interval[0] {
                result.push(new_interval.clone());
                result.extend_from_slice(&intervals[i..]);
                return result;
            } else {
                new_interval[0] = std::cmp::min(new_interval[0], interval[0]);
                new_interval[1] = std::cmp::max(new_interval[1], interval[1]);
            }
        }
        result.push(new_interval.clone());
        return result;
    }
}

fn main() {
    let mut intervals = vec![vec![1, 3], vec![6, 9]];
    let mut new_interval = vec![2, 5];
    let mut expected = vec![vec![1, 5], vec![6, 9]];
    test(intervals, new_interval, expected);

    intervals = vec![vec![1, 2], vec![3, 5], vec![6, 7], vec![8, 10], vec![12, 16]];
    new_interval = vec![4, 8];
    expected = vec![vec![1, 2], vec![3, 10], vec![12, 16]];
    test(intervals, new_interval, expected);

    intervals = vec![vec![1, 3], vec![6, 9]];
    new_interval = vec![3, 5];
    expected = vec![vec![1, 5], vec![6, 9]];
    test(intervals, new_interval, expected);

    intervals = vec![vec![1, 3], vec![6, 9]];
    new_interval = vec![9, 15];
    expected = vec![vec![1, 3], vec![6, 15]];
    test(intervals, new_interval, expected);

    intervals = vec![vec![1, 3], vec![6, 9]];
    new_interval = vec![11, 15];
    expected = vec![vec![1, 3], vec![6, 9], vec![11, 15]];
    test(intervals, new_interval, expected);
}

fn test(intervals: Vec<Vec<i32>>, new_interval: Vec<i32>, expected: Vec<Vec<i32>>) {
    let result = Solution::insert(intervals.clone(), new_interval.clone());

    println!("\nresult: ");
    array_colored(result.clone());
    println!("assert: {}", bool_colored(result == expected));
}

fn bool_colored(value: bool) -> ColoredString {
    let text;
    if value {
        text = value.to_string().green();
    } else {
        text = value.to_string().red();
    }
    return text;
}

fn array_colored(nums: Vec<Vec<i32>>) {
    nums.iter().for_each(|it| {
        println!("{:#?}", it);
    });
}
